Add support for default credentials.

orestica opened the pull request:
This pull request goes hand in hand with the following pull request: https://github.com/google/google-api-python-client/pull/4

Once these two pull requests are accepted and integrated into their respective repositories, a Google API call will be authenticated as simple as building the service like this:

service = build("compute", "v1", credential=GoogleCredential.get_default_credential())

This will work when the code is being run on a VM inside Google Compute Engine, a Google App Engine application or on any other off-fabric platform (in this last case, a credentials file needs to be put together containing either user or service account credentials and an environment variable pointed to it OR a 'gcloud auth login' command needs to be run before using the default credentials).

Commit:
Add support for GoogleCredential.GetDefaultCredential().

Commit:
Complete support for GoogleCredential.GetDefaultCredential().

Commit:
Add testing scripts for GoogleCredential.GetDefaultCredential().

Commit:
Remove a debugging print.

Commit:
Revert the googleapiclient.discovery changes (as they need to be done in
google-api-python-client repository).
Move the sample code in samples package.
Add dependency on rsa library for the oauth2client repository.

Commit:
Add support for the 'well known file'.

Commit:
Rename scopesRequired()->scopes_required() and
createdScoped()->create_scoped().

Commit:
Convert the methods from GetDefaultCredential() to
get_default_credential() format.

Add get_access_token() method to OAuth2Credentials class (effectively,
all credentials).

Add sign_blob() and get_service_account_name() methods to
ServiceAccountCredentials class.

Change how the private key (in pkcs8 format) is read inside
ServiceAccountCredentials class.

Add pyasn1 and pyasn1_modules as required libraries for the oauth2client
library.

Commit:
Add tests for ServiceAccountCredentials class.

Commit:
Fix the default_credential.get_well_known_file() method to return None
if there is no such file.

Commit:
Do not look for credentials in local files if running inside Google App
Engine.

Commit:
Merge branch 'master' of https://github.com/google/oauth2client into testing

Commit:
Add helpful comments for the PROJECT and ZONE constants in the sample
code.

Commit:
Bring the origin in sync with the upstream.

Commit:
Rename tests/test_oauth2client_jwt_rsa.py -> tests/test_jwt_rsa.py.

Commit:
Add version constraints for the newly required libraries: pyasn1 (0.1.7), pyasn1-modules (0.0.5) and rsa (3.1.4).

Commit:
Add the private key id to the JWT's header.

Commit:
Import HttpMockSequence from http_mock and comment out self.assertIsInstance() until I figure out how to make it work across all Python versions.

Commit:
Comment out self.assertIsInstance() until I figure out how to make it work across all Python versions.

Commit:
Give a more helpful error message in case a VM is created without a service account and an access token is requested from inside such a VM.

Commit:
Merge branch 'master' of https://github.com/orestica/oauth2client

Commit:
Reinforce the Google Compute Engine detection according to https://developers.google.com/compute/docs/metadata#runninggce .

craigcitro commented on oauth2client/default_credential.py:
```diff
@@ -0,0 +1,215 @@
+    elif env_name == 'GCE_PRODUCTION':
+      return get_default_credential_GCE(scopes)
+    else:
+      raise Exception('Either GOOGLE_CREDENTIALS_DEFAULT '
```
We want to raise exceptions from this package, not just base `Exception`

(sample comment)

Commit:
Rename credential parameter of the build method: build(credential) ->
build(credentials).

Commit:
Applied Anthony's suggestions from his review of my code (first round).

Commit:
Rename jwt.py -> service_account.py .

Commit:
Annotate Credentials.get_default() method with @staticmethod, instead of
@classmethod.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
 
+  @staticmethod
+  def get_default():
+    """Provides default credentials to be used in authenticating
```
If this is the starting point for the online docs, I would include a very breif summary what this does, i.e it uses the GAE or GCE built in service accounts, or it uses the environment variable or it uses the credential provided by "gcloud auth logic"

orestica replied:
I moved the comment to the newly created GoogleCredentials class' level. I put a link to https://developers.google.com/accounts/docs/default-credentials for further details.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+                      'accounts/docs/default-credential')
+
+
+def _get_google_credentials_default():
```
A nit: I found the name of this method confusing at first before i read the code to discover that it is environment variable path. Perhaps call it _get_environment_variable_credentials or something with environment_variable in it.

orestica replied:
I changed it to _get_environment_variable_file().

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+  """Get the well known file produced by command 'gcloud auth login'."""
+
+  WELL_KNOWN_CREDENTIALS_FILE = "credentials_default.json"
+  CLOUDSDK_CONFIG_WORD = "gcloud"
```
In the Java version, Brian Eaton suggested I change this to CLOUDSDK_CONFIG_DIRECTORY.

orestica replied:
Done.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+          client_credentials['client_secret'],
+          client_credentials['refresh_token'],
+          None,
+          'https://accounts.google.com/o/oauth2/token',
```
This must be used elsewhere in the file. Can it be a constant?

orestica replied:
Done.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+        _raise_exception_for_missing_fields(missing_fields)
+
+      return OAuth2Credentials(
+          None,
```
Minor: this call might be a good place to use named arguments as many of the arguments are not obvious from context.

orestica replied:
Done.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+          client_credentials['client_email'],
+          client_credentials['private_key_id'],
+          client_credentials['private_key'],
+          [])
```
Named argument just for the scopes might be more readable.

orestica replied:
Added for all of them, to be consistent.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+      if len(missing_fields) > 0:
+        _raise_exception_for_missing_fields(missing_fields)
+
+      return service_account.ServiceAccountCredentials(
```
In the Java version we do check for an pass in the "client_id" in addition to the email. The Java class does actually have inputs for it, although it does not seem to do anything interesting with them other than make them visible on the object. We should perhaps discuss this in the next meeting.

orestica replied:
I added the dummy client_id in the game.
Do you want me to add a getter on the ServiceAccountCredentials class for it?

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+          client_credentials['private_key'],
+          [])
+  except Exception as ex:
+    if google_credentials_default:
```
This feels like an implementation detail of the calling method bleeding into the interior method. It may be more maintainble to have this catch in the calling method. This also allows implementing from_stream if we decide that it needs to be public.

orestica replied:
You are spot on.
I moved it one level up.

anthmgoogle commented on oauth2client/service_account.py:
```diff
@@ -0,0 +1,130 @@
+  MAX_TOKEN_LIFETIME_SECS = 3600 # 1 hour in seconds
+
+  def __init__(self,
+      service_account_email,
```
As mentioned, we should discuss whether we shoul add service_account_id here.

orestica replied:
I already introduced it.
Let's discuss whether we need to remove it now. :)

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
     """
     return Credentials()
 
+  @staticmethod
```
As per the spec, I was expecting this on a new class in this same file called GoogleCredentials inherited from Credentials.

orestica replied:
Done. I made OAuth2Credentials inherit from GoogleCredentials, instead of Credentials. Hopefully this won't break the backwards compatibility.

anthmgoogle commented on samples/call_compute_service.py:
```diff
@@ -0,0 +1,16 @@
+PROJECT = "bamboo-machine-422"  # Provide your own GCE project here
+ZONE = "us-central1-a"          # Put here a zone which has some VMs
+
+service = build("compute", "v1", credentials=Credentials.get_default())
```
Expecting this to be GoogleCredentials.get_default()

orestica replied:
It is now.

anthmgoogle commented on oauth2client/service_account.py:
```diff
@@ -0,0 +1,130 @@
+from pyasn1_modules.rfc5208 import PrivateKeyInfo
+
+
+class ServiceAccountCredentials(AssertionCredentials):
```
It was expected that this class inherit from a new GoogleCredentials class. If AssertionCredentials can have that injected as a parent and that is not a breaking change in Python, that should do the trick.

orestica replied:
I introduced GoogleCredentials inheriting from Credentials and made OAuth2Credentials inherit from GoogleCredentials. As AssertionCredentials inherits from OAuth2Credentials, we should be good.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -242,6 +257,211 @@ def from_json(cls, s):
+    print response
+    </code>
+    """
+
```
The java implementation caches this in a static variable, and the non-functional requirements indicate that repeated calls should not need to recreate the instance, re-parse files, etc. This also requires simple locking on the fetch. This is non-trivial to test. I Java I did this by putting all the static state into an internal class only visible to the tests, and this also makes the tests thread-safe if run concurrently.

orestica replied:
I moved _env_name variable at the module level so it is static for all practical purposes.

anthmgoogle commented on tests/test_oauth2client.py:
```diff
@@ -100,12 +101,21 @@ def set(self, key, value, namespace=''):
 
```
Some potential gaps in the testing: it would be good to test the file format parsing, and exposing and testing _from_stream should enable this, as that part of the call would not touch the file system, environment variables, etc. Coverage of the error cases is important too.

There seems to be no coverage of the environment variable or well known file discovery paths, including errors. These are not straight-forward to test so, it might not be inconsistent with the depth of the testing that exists already to skip or defer these. In Java I did this with an internal class holding both the state and virtual methods to allow mocking things like checking for files or reading environment variables.

orestica replied:
I'll work tomorrow on adding more tests.
And once I do that, I'll push my code so you'll be able to see all the changes I've made at your suggestions.

orestica commented on tests/test_oauth2client.py:
```diff
@@ -526,7 +562,7 @@ def test_exchange_id_token_fail(self):
       'some random code', http=http)
 
-  def test_exchange_id_token_fail(self):
+  def test_exchange_id_token_fail_encore(self):
```
Please expand the code above.
There were two test methods with the same name: test_exchange_id_token_fail.
I thought I would be a good citizen if I differentiated them somehow.

craigcitro replied:
Yes! I also spotted this the other day and fixed it, but apparently hadn't pushed to this repo; it's now e48d951.

Commit:
Applied Anthony's suggestions from his review of my code (second round).
Added tests for the newly added class GoogleCredentials.

Commit:
Catch and properly handle the ValueError that might result from parsing
json.

Commit:
Parsing a malformed json seems to fail differently on different versions
of Python.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -169,6 +173,21 @@ def apply(self, headers):
+    """Whether this Credentials object is scopeless.
+    
+    create_scoped(scopes) method needs to be called in order to create
+    a proper Credentials object.
```
Perhaps "a Credentials object for API calls" instead of "a proper Credentials object".

orestica replied:
My intent was to convey the information that a Credentials object of the appropriate class would be returned.
Applied your suggestion.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -169,6 +173,21 @@ def apply(self, headers):
     """
     _abstract()
+    
+  def create_scoped_required(self):
```
The spec indicates that these should be surfaced on GoogleCredentials, not Credentials. We can discuss when we talk about the surface, but the equivalent for Java is on the Google-specific class.

orestica replied:
Moved the two new methods to GoogleCredentials class.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -243,6 +262,246 @@ def from_json(cls, s):
+      return 'UNKNOWN'
+
+
+_env_name = _get_environment()
```
This looks like it may have some side effects we might want to prevent. For example, would this do the server call now whenever the Credentials type was touched? It also looks like it might do this query  even if the environment variable or well known file were present. This could be prevented by doing the file checks first, doing this on-demand when get_default is called, and keeping a flag to make sure this query happens only once.

orestica replied:
I modified the code so _get_environment() gets called only once: the very first time the GoogleCredentials.get_default() is called.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -243,6 +262,246 @@ def from_json(cls, s):
+                                    'environment variable must be set or '
+                                    'you need to run "gcloud auth login"! '
+                                    'Please see https://developers.google.com/'
+                                    'accounts/docs/default-credential')
```
Note, I changed this url to be plural: default-credentials

orestica replied:
Good catch.
Done.

Commit:
Applied Anthony's suggestions from his review of my code (third round).
Made ServiceAccountCredentials class private.
Implemented the static method GoogleCredentials.from_stream().
Made GoogleCredentials class inherit from OAuth2Credentials class.

Commit:
Annotate GoogleCredentials' constructor with @util.positional(7).

Commit:
Annotate GoogleCredentials' constructor with @util.positional(8).

Commit:
Deannotate GoogleCredentials' constructor with @util.positional(8) and
name the revoke_uri parameter when calling the OAuth2Credentials'
constructor.

Commit:
Document the GoogleCredentials' constructor.

Commit:
Get rid of an unnecessary parameter.

Commit:
Refactor the GoogleCredentialsTests class by introducing the
GoogleCredentialsTests.get_a_google_credentials_object() method.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  PROJECT = 'bamboo-machine-422'  # replace this with one of your projects
+  ZONE = 'us-central1-a'          # replace this with the zone you care about
+
+  service = build('compute', 'v1', credentials=Credentials.get_default())
```
Should this get GoogleCredentials.get_default()

orestica replied:
Yes, it should. Nice catch.
Done.

anthmgoogle commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  def from_stream(credential_file):
+    """Create a Credentials object by reading the information from a given file.
+    
+    It returns an object of type _ServiceAccountCredentials/GoogleCredentials.
```
I would just say GoogleCredentials here, as the _ServiceAccountCredentials should probably not be referenced in public docs and it should inherit from GoogleCredentials.

orestica replied:
Done.

anthmgoogle commented on tests/test_oauth2client.py:
```diff
@@ -107,6 +120,314 @@ def test_to_from_json(self):
+    setattr(client, '_computed_env_name', False)
+
+  def tearDown(self):
+    self.reset_env('SERVER_SOFTWARE', self.env_server_software)
```
This pattern will have side-effects that could cause errors if the tests are run multi-threaded or if the there is an error in the tear-down code. This might not be a problem depending on how Python unit tests get run. It may be worth double-checking with Craig Citro/John Asmuth to see if it is OK to have tests that modify the process environment like this. If not, this can be handled by having the product code go through an non-public overridable virtual call for things like checking the environment.

orestica replied:
Hopefully, Craig will enlighten me how to do this (more) properly, once you give me your OK to get him involved.

craigcitro replied:
A few thoughts:
- The tests are never run in any sort of parallel/multithreaded environment, so there's unlikely to be a problem on that front.
- Most of the env stuff is harmless enough, but modifying anything in `os` is more worrisome.
- There's definitely a cleaner way to handle this in python, though I'm not sure how much you want to do. The "right" answer here is to write a [context manager](https://docs.python.org/2/library/contextlib.html); roughly speaking, you'd write something that looks like (where you're welcome to change the names):
  
  ```
  @contextlib.contextmanager
  def SetupEnv(desired_os, desired_server_software):
    original_server_software = os.environ.get('SERVER_SOFTWARE')
    os.environ['SERVER_SOFTWARE'] = desired_server_software
    ... <more env setup> ...
    yield
    os.environ['SERVER_SOFTWARE'] = original_server_software
    ... <more resetting> ...
  ```
  
  Then you can just say
  
  ```
  with SetupEnv('nt', 'DEV_APPSERVER'):
    ... <tests for that env> ...
  ```

Commit:
Applied Anthony's suggestions from his review of my code (fourth round).

Commit:
Merge branch 'master' of https://github.com/google/oauth2client

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  
+  The default credentials are being constructed as a function of the environment
+  where the code is being run. More details can be found on this page:
+  https://developers.google.com/accounts/docs/default-credentials .
```
Is there even a draft of this to look at somewhere?

orestica replied:
Yes. Here: https://docs.google.com/a/google.com/document/d/1jvTfUBqBgqVDQI24kzoE_YrAUbl6j4kKUshaSGWw0E0

craigcitro replied:
Feel free to drop the `.`, which is always awkward with URLs anyway.

craigcitro commented on oauth2client/appengine.py:
```diff
@@ -196,6 +197,13 @@ def _refresh(self, http_request):
     self.access_token = token
 
+  def create_scoped_required(self):
+    return not bool(self.scope)
```
You can drop the `bool` -- just `not self.scope` works (and is preferred by various style guides).

orestica replied:
Done. (also in gce.py and account_service.py)

craigcitro commented on oauth2client/appengine.py:
```diff
@@ -196,6 +197,13 @@ def _refresh(self, http_request):
+
+  def create_scoped(self, scopes):
+    return AppAssertionCredentials(scopes,
+                                   **self.kwargs)
```
bump this up one line.

orestica replied:
Done.

craigcitro commented on oauth2client/client.py:
```diff
@@ -99,6 +99,10 @@ class NonAsciiHeaderError(Error):
 
 
+class DefaultCredentialsError(Error):
+  """Error retrieving the default credentials."""
```
we seem to have a little inconsistency around how we talk about this -- the external docs seem to prefer the slightly unfortunate 'the Default Credentials', so we should probably go with that.

orestica replied:
Done. (also in other 5 places in this file)

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+      if http_request is None:
+        http_request = httplib2.Http().request
+      response, _ = http_request('http://metadata.google.internal')
+      if ('metadata-flavor' in response and
```
Python dictionaries have a `.get()` method that simplifies this:

```
if response.get('metadata-flavor') == 'Google':
```

orestica replied:
Indeed, I should have used the get('metadata-flavor', '') instead of checking like this.
In the new implementation (using urllib2 instead of httplib2), it is not needed anymore.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+def _get_environment(http_request=None):
+  """Detect the environment the code is being run on."""
+
+  global _computed_env_name
```
Why the second variable? Seems easier to just check for `_env_name` being `None`. (If the thought was "oh performance", remember it's Python.)

orestica replied:
You are right. It's not worth explaining why I chose to use both (it has to do with me learning Python).
I got rid of the _computed_env_name variable.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  elif server_software.startswith('Development/'):
+    _env_name = 'GAE_LOCAL'
+  else:
+    import httplib2
```
I'd prefer to use `urllib2` in this case, for three reasons:
- First and foremost, I eventually want to drop the dependency on `httplib2` in this library. No point in adding more references.
- As a symptom of why I want to drop it, last week the GCS folks hit some strange issues with `httplib2` in this particular codepath (in apitools, but it's doing the same thing). We didn't dig much, but switching to `urllib2` fixed it.
- Given that we're testing for GCE anyway, I don't see an important use case for the user passing in a custom http instance.

craigcitro replied:
[Sample code](https://github.com/GoogleCloudPlatform/gsutil/blob/master/gslib/third_party/storage_apitools/credentials_lib.py#L135) where we use `urllib2` in `gsutil`, for exactly this purpose.

orestica replied:
Done.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+
+
+class GoogleCredentials(OAuth2Credentials):
+  """Provides default credentials to be used in authenticating
```
This should be one line -- I'd say just `Default credentials for use in calling Google APIs.`

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  Here is an example of how to use the default credentials for a service that
+  requires authentication:
+
+  <code>
```
I'd rather see this in the samples and have a pointer here; multiple copies of hand-written docs just get out of sync too fast.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  service = build('compute', 'v1', credentials=GoogleCredentials.get_default())
+
+  resource = service.instances()
+  request = resource.list(project=PROJECT, zone=ZONE)
```
I'd at least combine this line with the one above.

orestica replied:
Combined the previous two lines into one.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  print response
+  </code>
+
+  Exceptions:
```
This probably belongs on the docstrings of the method(s) that raise it, not here.

orestica replied:
I borrowed the idea from the docstring of class AccessTokenCredentials.
I moved it down to get_default() and from_stream() methods.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+    DefaultCredentialsError: raised when the credentials fail to be retrieved.
+  """
+
+  def __init__(self,
```
Let's pack these a little tighter.

orestica replied:
Done.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+    """Create an instance of GoogleCredentials.
+
+    This constructor is not usually called by the user, instead
+    GoogleCredentials objects are instantiated by the
```
Drop "the".

orestica replied:
Done.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+
+    This constructor is not usually called by the user, instead
+    GoogleCredentials objects are instantiated by the
+    GoogleCredentials.from_stream() and GoogleCredentials.get_default().
```
silly: and -> or

orestica replied:
Done.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+
+  @staticmethod
+  def get_default():
+    """Get the default credentials appropriate for the environment in which the
```
This needs to be a one-liner. Maybe `Fetch the default credentials for this environment`?

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+                                    "credentials for details.")
+
+  @staticmethod
+  def from_stream(credential_file):
```
This isn't a file, it's a filename.

Relatedly, this would be significantly more useful if it took a file instead of a filename, but that would involve breaking compatibility with the other clients, possibly.

orestica replied:
I changed it to credential_filename. Let's leave like this.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+      try:
+        return _get_default_credential_from_file(credential_file)
+      except (DefaultCredentialsError, ValueError) as error:
+        helping_message = ' (provided as parameter to the from_stream() method)'
```
maybe `helping` -> `help`?

craigcitro replied:
Or `extra_help`?

orestica replied:
Renamed it to extra_help.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+    elif _env_name == 'GCE_PRODUCTION':
+      return _get_default_credential_GCE()
+    else:
+      raise DefaultCredentialsError("The Default Credentials are not available."
```
We should use `'` for consistency here; in addition, I'd vote to start this on the line below, so that we don't use up so much vertical space.

orestica replied:
I used " because there is 'gcloud auth login' somewhere there inside the string.
I consolidated the text a little bit.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+
+
+def _get_well_known_file():
+  """Get the well known file produced by command 'gcloud auth login'."""
```
I have two worries about all these `gcloud` references:
- `gcloud` doesn't cover nearly all the Google APIs (eg Drive); I don't want anyone to think this library isn't useful outside cloud APIs.
- If the command name changes, we have a **lot** of references to update.

craigcitro replied:
And now, reading the body of this function, we're taking quite a dependency on logic about where `gcloud` puts its credentials. Is there a way to depend on this logic over there? (Maybe they could set an env var?)

orestica replied:
We plan to announce the 'gcloud auth login' in the public go to page for education purposes.
I sent you an internal link (via email) regarding where I borrowed the code from.

craigcitro replied:
OK; I'd ultimately like to figure out a way to have gcloud tell us this file location, but that doesn't have to happen right now.

orestica replied:
I left a TODO.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+    except KeyError:
+      # This should never happen unless someone is really messing with things.
+      drive = os.environ.get('SystemDrive', 'C:')
+      default_config_path = os.path.join(drive,
```
Will this all fit on one line?

orestica replied:
Yes.
I moved everything on a single line.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  if os.path.isfile(default_config_path):
+    return default_config_path
+  else:
+    return None
```
You don't need this -- falling off the end of a function returns `None`.

orestica replied:
Got rid of it.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  with open(default_credential_file) as default_credential:
+    client_credentials = service_account.simplejson.load(default_credential)
+
+  if ('type' not in client_credentials or
```
```
if client_credentials.get('type') not in ('authorized_user', 'service_account'):
```

For the record, just depending on strings is going to get us into trouble one day. Let's not worry about that here.

orestica replied:
Thanks.
Python style is more concise, indeed.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+  if ('type' not in client_credentials or
+      client_credentials['type'] != 'authorized_user' and
+      client_credentials['type'] != 'service_account'):
+    raise DefaultCredentialsError("'type' field should be defined "
```
This is a place where `"` is totally legit (because it's a string containing `'`).

orestica replied:
This was my rationale behind the previous ", as well.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+    raise DefaultCredentialsError("'type' field should be defined "
+                                  "(and have one of the 'authorized_user' "
+                                  "or 'service_account' values)")
+
```
Two thoughts:
- Lots of duplication here; any reason we can't collapse it?
- Should some/all of this code live on the classes for the credentials they create?

orestica replied:
I restructured it a little bit.
I would rather keep the validation of fields' existence here, than pushing it down into the credentials classes.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+                                  "or 'service_account' values)")
+
+  if client_credentials['type'] == 'authorized_user':
+    missing_fields = []
```
Kudos on the good error message here!

Python has a pretty featureful `set` type; I think it'd help here. For instance:

```
required_fields = set(['client_id', 'client_secret', 'refresh_token'])
missing_fields = required_fields.difference(client_credentials.keys())
if missing_fields:
  ...
```

orestica replied:
This is more elegant, indeed. Thank you.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+                                helping_message + ': ' + str(error))
+
+
+def _get_default_credential_GAE():
```
Seems like this could live on the GAE credential class?

orestica replied:
Moving it as a public method to appengine.AppAssertionCredentials class would increase the new surface we introduce. We don't want that.
Moving it as a private method over there wouldn't gain us much: we'd still need an import followed by a one line to call this method.
I would rather let it here.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,337 @@ def _revoke(self, http_request):
+def _get_default_credential_GAE():
+  from oauth2client.appengine import AppAssertionCredentials
+
+  return AppAssertionCredentials([])
```
it really doesn't need any args?

orestica replied:
It needs only scopes. Passing an empty list of scopes defers them for later filling.

craigcitro commented on oauth2client/gce.py:
```diff
@@ -87,4 +88,14 @@ def _refresh(self, http_request):
         raise AccessTokenRefreshError(str(e))
       self.access_token = d['accessToken']
     else:
+      if content.find('Error 404') >= 0:
```
Can't we check against the status code on the response? I don't like grepping error strings unless we have to.

orestica replied:
Done.

craigcitro commented on oauth2client/service_account.py:
```diff
@@ -0,0 +1,132 @@
+This credentials class is implemented on top of rsa library.
+"""
+
+__author__ = 'orest@google.com (Orest Bolohan)'
```
We don't actually need `__author__` tags anymore.

orestica replied:
Dropped it.

craigcitro commented on oauth2client/service_account.py:
```diff
@@ -0,0 +1,132 @@
+
+  MAX_TOKEN_LIFETIME_SECS = 3600 # 1 hour in seconds
+
+  def __init__(self,
```
Let's pack these arguments.

orestica replied:
Done.

craigcitro commented on oauth2client/service_account.py:
```diff
@@ -0,0 +1,132 @@
+    return (self._private_key_id,
+            rsa.pkcs1.sign(blob, self._private_key, 'SHA-256'))
+
+  def get_service_account_email(self):
```
Just make this

```
@property
def service_account_email(self):
  return self._service_account_email
```

and then you have a getter; callers can just say `credential.service_account_email` and it does what you expect.

orestica replied:
Neat.
Thank you.

craigcitro commented on oauth2client/service_account.py:
```diff
@@ -0,0 +1,132 @@
+      simplejson.dumps(data, separators = (',', ':'))\
+          .encode('UTF-8')).rstrip('=')
+
+def _get_pk(private_key_pkcs8_text):
```
let's expand out the name (pk = private key?)

orestica replied:
Done.

craigcitro commented on samples/app.yaml:
```diff
@@ -0,0 +1,10 @@
+application: bamboo-machine-422
```
actually, why don't we make a folder for the sample GAE app inside `samples/`? it's always frustrating when you don't know which file(s) are part of a bundle like that.

orestica replied:
Done.

craigcitro commented on samples/call_compute_service_from_gae.py:
```diff
@@ -0,0 +1,24 @@
+  def get(self):
+    self.response.write(get_instances())
+
+app = webapp2.WSGIApplication([
```
one line

orestica replied:
Done.

craigcitro commented on setup.py:
```diff
@@ -26,6 +26,9 @@
 install_requires = [
     'httplib2>=0.8',
+    'pyasn1==0.1.7',
+    'pyasn1_modules==0.0.5',
```
do we know anything about these modules? taking a dependency on something at version `0.0.5` worries me; have we at least forked them ourselves in case they disappear?

orestica replied:
Should I get rid of the versions?

craigcitro replied:
I'm less worried about the versions, and more worried about the prospect of those packages disappearing from the world.

orestica replied:
Replied to you on email.

craigcitro commented on tests/data/publickey_openssl.pem:
```diff
@@ -0,0 +1,9 @@
+-----BEGIN PUBLIC KEY-----
```
is this a public key you have the private key for?

orestica replied:
This public key was generated from the certificate from the existing publickey.pem file, by running:

openssl x509 -pubkey -noout -in publickey.pem

This public key goes hand in hand with the private key from various privatekey files from the same directory (as you'll see in test_service_account.py).

craigcitro replied:
aha -- I forgot there was already a key in this repo. no worries.

orestica replied:
Ack.

Commit:
Started applying Craig's suggestions.

craigcitro commented on tests/test_oauth2client.py:
```diff
@@ -107,6 +120,314 @@ def test_to_from_json(self):
+    self.reset_env('APPDATA', self.env_appdata)
+    os.name = self.os_name
+
+  def reset_env(self, env, value):
```
I think you could simplify this using `os.environ.pop`, but it's not deeply important.

orestica replied:
Is this what you had in mind?

craigcitro replied:
Yep. Can you add a docstring for this function?

orestica replied:
Done.

craigcitro commented on tests/test_oauth2client.py:
```diff
@@ -107,6 +120,314 @@ def test_to_from_json(self):
+    self.assertEqual('Python client library', credentials.user_agent)
+
+  def get_a_google_credentials_object(self):
+    return GoogleCredentials(None,
```
Let's compact these

orestica replied:
Done.

craigcitro commented on tests/test_oauth2client.py:
```diff
@@ -107,6 +120,314 @@ def test_to_from_json(self):
+                     credentials.create_scoped(['dummy_scope']))
+
+  def test_get_environment_gae_production(self):
+    os.environ['SERVER_SOFTWARE'] = 'Google App Engine/XYZ'
```
Probably want to reset this after.

craigcitro replied:
(and below)

orestica replied:
setUp() and tearDown() take care of this.

craigcitro commented on tests/test_oauth2client.py:
```diff
@@ -107,6 +120,314 @@ def test_to_from_json(self):
+  def test_get_environment_variable_file_error(self):
+    nonexistent_file = datafile('nonexistent')
+    os.environ['GOOGLE_CREDENTIALS_DEFAULT'] = nonexistent_file
+    try:
```
You should check out `self.assertRaises`, which will let you clean this up.

orestica replied:
self.assertRaises() looks nice if I just want to make sure a DefaultCredentialsError exception is being raised. But could I test its message the way I do it here?

craigcitro replied:
Yes -- also check out `assertRaisesRegexp`.

orestica replied:
Done.
Thank you.

craigcitro commented on tests/test_service_account.py:
```diff
@@ -0,0 +1,121 @@
+from oauth2client.service_account import _ServiceAccountCredentials
+
+
+def datafile(filename):
```
The right way to do this is with `pkgutil.get_data`; we do this in the [bq code](https://code.google.com/p/google-bigquery-tools/source/browse/bq/bigquery_client.py#471), but it also requires a little work in `setup.py` to make sure you've got it right. You can leave a `TODO` on this one if you'd like.

orestica replied:
I got inspired from test_jwt.py, which uses this very pattern.
I put a TODO.

craigcitro replied:
TODO works.

Commit:
Continued addressing Craig's suggestions.

Commit:
Pull from upstream.

Commit:
Another round of applying Craig's suggestions.

craigcitro commented on oauth2client/appengine.py:
```diff
@@ -164,6 +164,7 @@ def __init__(self, scope, **kwargs):
           unspecified, the default service account for the app is used.
     """
     self.scope = util.scopes_to_string(scope)
+    self.kwargs = kwargs
```
Any reason to leave this public?

(I could _swear_ I already made this comment, but don't see it; if I'm repeating myself, just tell me I'm getting senile.)

orestica replied:
No reason. I just wanted to preserve the look of the surrounding instance variable.
I made it private.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,304 @@ def _revoke(self, http_request):
+      if urllib2_urlopen is None:
+        urllib2_urlopen = urllib2.urlopen
+      response = urllib2_urlopen('http://metadata.google.internal')
+      if 'Metadata-Flavor: Google\r\n' in response.info().headers:
```
It might be worth calling `.rstrip()` on the headers and checking without the `\r\n` -- that seems like it might fail painfully one day.

craigcitro replied:
or, alternately:

```
if any('Metadata-Flavor: Google' in h for h in response.info().headers):
```

orestica replied:
Chose the 'if 'Metadata-Flavor: Google\r\n' in response.info().headers:' option.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,304 @@ def _revoke(self, http_request):
+
+
+class GoogleCredentials(OAuth2Credentials):
+  """Provides Default Credentials to be used in authenticating
```
This still needs to be shortened to one line.

orestica replied:
I remember seeing this suggestion in the email, but it did not show up in https://github.com/google/oauth2client/pull/6/files
Have you experienced inconsistencies with github's reviewing tools?

Done.

craigcitro replied:
Yes, I think sometimes comments are hard to find unless you trawl through all the commits. It's sub-par.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,304 @@ def _revoke(self, http_request):
+
+  if client_credentials['type'] == 'authorized_user':
+    required_fields = set(['client_id', 'client_secret', 'refresh_token'])
+  else: # client_credentials['type'] == 'service_account'
```
nit: python-style wants two spaces before the comment (so `else:  # ...`).

craigcitro replied:
also, in this case, why not just combine this with the previous if:

```
credentials_type = client_credentials.get('type')
if credentials_type == 'authorized_user':
  ...
elif credentials_type == 'service_account':
  ...
else:
  raise DefaultCredentialsError(...)
```

orestica replied:
Done.

craigcitro commented on oauth2client/client.py:
```diff
@@ -820,7 +838,304 @@ def _revoke(self, http_request):
+  with open(default_credential_filename) as default_credential:
+    client_credentials = service_account.simplejson.load(default_credential)
+
+  if (client_credentials.get('type') not in ('authorized_user',
```
can you give these two strings (`'authorized_user'` and `'service_account'`) names that we use throughout?

orestica replied:
Done.

Commit:
Applied more suggestions from Craig.

Commit:
Drop using assertRaisesRegexp(), as it is not supported in Python 2.6 .

Commit:
Introduce the GOOGLE_CREDENTIALS_DEFAULT constant.
